File: problems/10_backtracking/0051_n_queens/solution.swift
// Problem: 0051. N-Queens
// Difficulty: Hard

import Foundation

class Solution {
    func solveNQueens(_ n: Int) -> [[String]] {
        var out: [[String]] = []
        var cols = Set<Int>(), d1 = Set<Int>(), d2 = Set<Int>()
        var placed: [Int] = []
        func dfs(_ r: Int) {
            if r == n {
                var board: [String] = []
                for c in placed {
                    var row = [Character](repeating: ".", count: n)
                    row[c] = "Q"
                    board.append(String(row))
                }
                out.append(board)
                return
            }
            for c in 0..<n {
                if cols.contains(c) || d1.contains(r - c) || d2.contains(r + c) { continue }
                cols.insert(c); d1.insert(r - c); d2.insert(r + c)
                placed.append(c)
                dfs(r + 1)
                placed.removeLast()
                cols.remove(c); d1.remove(r - c); d2.remove(r + c)
            }
        }
        dfs(0)
        return out
    }
}

func eqBoards(_ a: [[String]], _ e: [[String]]) -> Bool {
    if a.count != e.count { return false }
    let sa = a.map { $0.joined(separator: "|") }.sorted()
    let sb = e.map { $0.joined(separator: "|") }.sorted()
    return sa == sb
}

let sol = Solution()
if !eqBoards(sol.solveNQueens(1), [["Q"]]) { fatalError("a") }
let r4 = sol.solveNQueens(4)
if r4.count != 2 { fatalError("b1") }
if !eqBoards(r4, [
    [".Q..", "...Q", "Q...", "..Q."],
    ["..Q.", "Q...", "...Q", ".Q.."],
]) { fatalError("b2") }
if sol.solveNQueens(5).count != 10 { fatalError("c") }
if sol.solveNQueens(6).count != 4 { fatalError("d") }
if !sol.solveNQueens(2).isEmpty { fatalError("e") }
print("All tests passed!")

File: problems/10_backtracking/0051_n_queens/solution.dart
// Problem: 0051. N-Queens
// Difficulty: Hard

List<List<String>> solveNQueens(int n) {
  final out = <List<String>>[];
  final cols = <int>{}, d1 = <int>{}, d2 = <int>{};
  final placed = <int>[];
  void dfs(int r) {
    if (r == n) {
      final board = <String>[];
      for (final c in placed) {
        final row = List<String>.filled(n, '.');
        row[c] = 'Q';
        board.add(row.join());
      }
      out.add(board);
      return;
    }
    for (var c = 0; c < n; c++) {
      if (cols.contains(c) || d1.contains(r - c) || d2.contains(r + c)) continue;
      cols.add(c); d1.add(r - c); d2.add(r + c);
      placed.add(c);
      dfs(r + 1);
      placed.removeLast();
      cols.remove(c); d1.remove(r - c); d2.remove(r + c);
    }
  }
  dfs(0);
  return out;
}

bool _eqBoards(List<List<String>> a, List<List<String>> b) {
  if (a.length != b.length) return false;
  String key(List<String> bd) => bd.join('|');
  final sa = a.map(key).toList()..sort();
  final sb = b.map(key).toList()..sort();
  for (var i = 0; i < sa.length; i++) if (sa[i] != sb[i]) return false;
  return true;
}

void main() {
  if (!_eqBoards(solveNQueens(1), [["Q"]])) throw 'a';
  final r4 = solveNQueens(4);
  if (r4.length != 2) throw 'b1';
  if (!_eqBoards(r4, [
    [".Q..", "...Q", "Q...", "..Q."],
    ["..Q.", "Q...", "...Q", ".Q.."],
  ])) throw 'b2';
  if (solveNQueens(5).length != 10) throw 'c';
  if (solveNQueens(6).length != 4) throw 'd';
  if (solveNQueens(2).isNotEmpty) throw 'e';
  print('All tests passed!');
}
